mono/utils/mono-merp.c

```diff
@@ -0,0 +1,493 @@
+	const char *home = g_get_home_dir ();
+	char *merpParamPath = g_strdup_printf ("%s/Library/Group Containers/UBF8T346G9.ms/MERP.Params.txt", home);
+	write_file (str, merpParamPath);
+	g_free (merpParamPath);
+
+	mach_port_t merpPort = MACH_PORT_NULL;
+
+	// Start merpGui application
+	// Assign to merpPort the port that merpGui opens
+	// Connecting to the service with name serviceName 
+	// that merpGUI starts
+	connect_to_merp (serviceName, &merpPort);
+
+	// Send this mach task to MERP over the port
+	send_mach_message (&merpPort);
+
+	// After we resume from the suspend, we are done
+	// We're spinning down the process shortly.
+	// This thread exits, and the crashing thread's
+	// waitpid on this thread ends.
+	return;
+}
+
+static void
+mono_init_merp (const char *serviceName, const char *signal, pid_t crashed_pid, intptr_t thread_pointer, MERPStruct *merp, MonoError *error)
+{
+	error_init (error);
+
+	g_assert (mono_merp_enabled ());
+
+	merp->pidArg = crashed_pid;
+	merp->archArg = get_merp_arch ();
+	merp->capabilitiesArg = 0x0;
+
+	merp->threadArg = thread_pointer;
+
+	// FIXME: time the runtime?
+	merp->timeArg = 0x0;
+
+	merp->exceptionArg = parse_exception_type (signal);
+
+	// If these aren't set, icall wasn't made
+	// don't do merp? / don't set the variable to use merp;
+	merp->bundleIDArg = config.appBundleID;
+	merp->signatureArg = config.appSignature;
+	merp->versionArg = config.appVersion;
+
+	// FIXME: Do we want these?
+	merp->devRegionArg = NULL;
+	merp->uiLidArg = NULL;
+
+	merp->serviceNameArg = serviceName;
+
+	// FIXME: Do we want these?
+	merp->appLoggerName = NULL;
+	merp->appLogSessionUUID = NULL;
+
+	// Should be set for our usage as per document
+	merp->isOfficeApplication = TRUE;
+	merp->isObjCException = FALSE;
+
+	mono_arch_memory_info (&merp->memRes, &merp->memVirt);
+
+	// No sessions right now
+	merp->treSessionID = NULL;
+	merp->exceptionCodeArg = NULL;
+	merp->exceptionAddressArg = NULL;
+
+	// Not certain? Maybe expose to config options
+	merp->isAppRegisteredWithMAU = TRUE;
+}
+
+void
+mono_merp_invoke (pid_t crashed_pid, intptr_t thread_pointer, const char *signal)
+{
+	// This unique service name is used to communicate with merp over mach service ports
+	char *serviceName = g_strdup_printf ("com.mono.merp.%.8x", crashed_pid);
+
+	ERROR_DECL (error);
+	MERPStruct merp = {0};
+	mono_init_merp (serviceName, signal, crashed_pid, thread_pointer, &merp, error);
+	mono_error_assert_ok (error);
+
+	GString *output = g_string_new ("");
+	mono_encode_merp (output, &merp);
+
+	if (config.log)
+		fprintf (stderr, "Results: \n(%s)\n", output->str);
+
+	// We send the merp over the port
+	mono_merp_send (output, serviceName);
+	g_string_free (output, TRUE);
+}
+
+void
+mono_merp_disable (void)
+{
+	memset ((void *)&config, 0, sizeof (MerpOptions));
+}
+
+void
+mono_merp_enable (const char *appBundleID, const char *appSignature, const char *appVersion, const char *merpGUIPath)
+{
+	config.appBundleID = appBundleID;
```

Comment:
Add a `initialized` variable to avoid initializing twice, or just use `config.enalble_merp`; it doesn't need to be threadsafe. You should also not assume the lifetime of these strings, and just `g_strndup` them.